### PR TITLE
Add more safe removal of weak listeners

### PR DESCRIPTION
**Pull request by robskillington**

### CODE REVIEW

**Comment by robskillington** [comment]
Okay so the reason this is an issue is because someone might call `removeListener(self)` as part of a signal firing and then `removeAtIndex` math is wrong because there is one less in the array already than previously though.

File: Signals/Signal.swift [hunk 1]
```diff
@@ -98,16 +98,20 @@ public class Signal<T> {
...
                 if !signalListener.dispatch(data) {
-                    signalListeners.removeAtIndex(index--)
+                    var hash = (signalListener as AnyObject).hash
+                    removeListeners[hash] = true
                 }
-                index++
             }
         }
+        
+        signalListeners = signalListeners.filter { signalListener in
```
Comment (artman) [inline Signals/Signal.swift]: Filters are really slow unfortunately. Should use something else. Let me check this branch out, write a test and check if I can make this faster.

Comment (robskillington) [inline Signals/Signal.swift]: Sounds good

Comment (artman) [inline Signals/Signal.swift]: But yeah, otherwise looks good, merging

**Pull request merged**